examples/some-issuer/src/bin/telegram.rs

```diff
@@ -0,0 +1,296 @@
+        help = "Address of the registry smart contract.",
+        env = "TELEGRAM_ISSUER_REGISTRY_ADDRESS"
+    )]
+    registry: ContractAddress,
+    #[clap(
+        long = "network",
+        help = "The network of the issuer.",
+        default_value = "testnet",
+        env = "TELEGRAM_ISSUER_NETWORK"
+    )]
+    network: Network,
+    #[clap(
+        long = "wallet",
+        help = "Path to the wallet keys.",
+        env = "TELEGRAM_ISSUER_WALLET"
+    )]
+    wallet: PathBuf,
+    #[clap(
+        long = "issuer-key",
+        help = "Path to the issuer's key, used to sign commitments.",
+        env = "TELEGRAM_ISSUER_KEY"
+    )]
+    issuer_key: PathBuf,
+    #[clap(
+        long = "max-register-energy",
+        help = "The amount of energy to allow for execution of the register credential \
+                transaction. This must be less than max block energy of the chain the service is \
+                connected to.",
+        default_value = "10000",
+        env = "TELEGRAM_ISSUER_MAX_REGISTER_ENERGY"
+    )]
+    max_register_energy: Energy,
+    #[clap(
+        long = "telegram-token",
+        help = "Bot token for Telegram.",
+        env = "TELEGRAM_BOT_TOKEN"
+    )]
+    telegram_bot_token: String,
+    #[clap(
+        long = "listen-address",
+        help = "Socket addres for the Telegram issuer.",
+        default_value = "0.0.0.0:8080",
+        env = "TELEGRAM_ISSUER_LISTEN_ADDRESS"
+    )]
+    listen_address: SocketAddr,
+    #[clap(
+        long = "url",
+        help = "URL of the Telegram issuer.",
+        default_value = "http://127.0.0.1:8080/",
+        env = "TELEGRAM_ISSUER_URL"
+    )]
+    url: Url,
+    #[clap(
+        long = "dapp-domain",
+        help = "The domain of the dApp, used for CORS.",
+        default_value = "http://127.0.0.1",
+        env = "TELEGRAM_ISSUER_DAPP_URL"
+    )]
+    dapp_domain: String,
+}
+
+#[derive(Clone)]
+struct AppState {
+    issuer: IssuerState,
+    telegram_bot_token: Arc<String>,
+}
+
+#[derive(Deserialize, Debug)]
+struct User {
+    id: u64,
+    first_name: String,
+    last_name: Option<String>,
+    username: Option<String>,
+    photo_url: Option<String>,
+    auth_date: u64,
+    hash: String,
+}
+
+type HmacSha256 = Hmac<Sha256>;
+impl User {
+    fn check(&self, telegram_bot_token: &str) -> anyhow::Result<()> {
+        let mut hasher = Sha256::new();
+        hasher.update(telegram_bot_token);
+        let key = hasher.finalize();
+
+        let mut mac = HmacSha256::new_from_slice(&key).unwrap();
+        mac.update(self.to_string().as_bytes());
+
+        let mac_bytes = mac.finalize().into_bytes();
+        let expected = hex::decode(&self.hash)?;
+
+        anyhow::ensure!(&mac_bytes[..] == &expected[..], "MAC did not match hash.");
+        Ok(())
+    }
+}
+
+/// Note: the exact format of this is important, as it is used for verification.
+/// See https://core.telegram.org/widgets/login
+impl Display for User {
```

Comment:
It seems odd to use a `Display` implementation for such a sensitive thing.

It would be better to use a dedicated function. Display usually does not have such strict requirements.